### ISSUE TITLE
Jobs are not processed but marked as successful

### ISSUE THREAD

**Issue by pkunze** [HangfireIO/Hangfire#1846]
We are facing a strange and terrifying problem.

At our QA environments, jobs are enqueued correctly. They are marked as processing and successful but obviously never actually ran.
This seems to happen after DB-Migrations (EF Core). 

![image](https://user-images.githubusercontent.com/1416798/114850107-b92e8180-9de0-11eb-9ea0-ba62821feb1d.png)

In the above image, 10 ms is way to fast for that job to run. It does not contain any catch clause that would swallow exceptions and all other Jobs are affected by the issue as well. Not a single job works (not even just sending an email).

We tried many things but are out of ideas.  Since we do not know the cause of the issue, we were not able to reproduce this locally but we have QA-Environments where this problem persists.

It seems to be database-related since if we drop the database and deploy/migrate from scratch the jobs are being picked up just fine. Obviously, this is no option vor Production :)
We think this happens after running a database migration (in Startup.Configure()).

Hangfire is configured according to the recommended settings in the docs.
We are using SQL Server Storage.

Is there any diagnostics I could be doing to determine/fix the cause of this issue?

What we tried so far: 
- making the job synchronous (removing `async`)
- Logging
- Upgrading Hangfire to the Latest version
- Redeploy / Restarting the app without DB drop
- running only a single hangfire server

**Comment by odinserj** [comment]
Ensure your method's signature doesn't look like `public async void MyMethod`, and looks like `public async Task MyMethod` – otherwise Hangfire will be unable await the asynchronous task. If everything is correct, please ensure that every async call is awaited in your method.

**Comment by pkunze** [comment]
Good point! 
Actually we have Analyzers checking both of those potential problems and I just checked and made sure that every `async` Method is `await`ed and that we don't use `async void` Method Signatures.

I even made Jobs kompletely synchronous (void, no async calls) but that didn't help either.

Also, this doesn't explain why jobs run after wiping the Database. Which is why we suspect Migrations to be the root cause.

**Comment by odinserj** [comment]
> I even made Jobs kompletely synchronous (void, no async calls) but that didn't help either.

It is possible to run an asynchronous method from a synchronous one, and in this case the call will not be awaited, giving the results you've described. So I can't say anything without source code for your method.

```csharp
public void MyMethod()
{
    SomeApi.RunSomeProcess(); 
}

// Somewhere in a third-party library
public async Task RunSomeProcess() { /**/ }
```

If you are talking about migrations for Hangfire.SqlServer, then they can't cause this without bringing the whole processing down. So source code is required for the method, and optionally Hangfire.Core and Hangfire.SqlServer versions and their configuration logic.

**Comment by pkunze** [comment]
> It is possible to run an asynchronous method from a synchronous one, and in this case the call will not be awaited, giving the results you've described

You're right. The mayority of Jobs is async right now (besides sending E-Mail). We just tried to make a job synchronous (replacing all Async Method calls by their sync. Overloads/Counterparts along the way) to see if this fixes the issue (it did not). We then noticed this doesn't fix the Issue and the (sync) E-Mail Job is also affected, **unit we wipe the Database and recreate it from scratch** (with our schema and hangfires).

Here is the most simple Job I can show:

```csharp
public class EmailService : IEmailService
{
	private readonly EmailConfig ec;

	public EmailService(IOptions<EmailConfig> emailConfig)
	{
		this.ec = emailConfig.Value;
	}

	public void SendEmail(string toAddress, string subject, string message)
	{
		this.SendEmail(toAddress, subject, message, new Dictionary<string, string>());
	}

	public void SendEmail(string toAddress, string subject, string message, Dictionary<string, string> mailHeaders)
	{
		var emailMessage = new MimeMessage();

		emailMessage.From.Add(new MailboxAddress(this.ec.FromName, this.ec.FromAddress));
		emailMessage.To.Add(new MailboxAddress(string.Empty, toAddress));
		emailMessage.Subject = subject;
		emailMessage.Body = new TextPart(TextFormat.Html) { Text = message };

		foreach (var (key, value) in mailHeaders)
		{
			emailMessage.Headers.Add(key, value);
		}

		using var client = new SmtpClient();

		if (this.ec.TrustAllCertificates)
		{
			client.ServerCertificateValidationCallback = (s, c, h, e) => true;  // invalides Zertifikat akzeptieren
		}

		client.Connect(
			this.ec.MailServerAddress,
			Convert.ToInt32(this.ec.MailServerPort),
			false);

		client.Send(emailMessage);

		client.Disconnect(true);
	}
}
```

I just checked that none of the Methods used there are `async` lacking the `Async`-Method Suffix:
![image](https://user-images.githubusercontent.com/1416798/114874663-d2ddc200-9dfc-11eb-8c84-25576e02f016.png)
![image](https://user-images.githubusercontent.com/1416798/114874731-e426ce80-9dfc-11eb-872f-131be89a0742.png)
![image](https://user-images.githubusercontent.com/1416798/114874768-ec7f0980-9dfc-11eb-9c62-78212721a289.png)

Our Hangfire Server/Dashboards are configured according to the Docs:

```csharp

// in Startup.Configure()
services.AddHangfire(configuration => configuration
    .SetDataCompatibilityLevel(CompatibilityLevel.Version_170)
    .UseSimpleAssemblyNameTypeSerializer()
    .UseRecommendedSerializerSettings()
    .UseNLogLogProvider()
    .UseSqlServerStorage(connectionString, new SqlServerStorageOptions
    {
        CommandBatchMaxTimeout = TimeSpan.FromMinutes(5),
        SlidingInvisibilityTimeout = TimeSpan.FromMinutes(5),
        QueuePollInterval = this.CurrentEnvironment.IsDevelopment() ? TimeSpan.FromSeconds(20) : TimeSpan.Zero,
        UseRecommendedIsolationLevel = true,
        UsePageLocksOnDequeue = true,
        DisableGlobalLocks = true,
        EnableHeavyMigrations = true
    }));

// ...

// in Startup.ConfigureServices()
app.UseHangfireDashboard(options: new DashboardOptions
{
    DashboardTitle = "Übersicht Background-Jobs",
    Authorization = new[] { new HangfireAuthorizationFilter() }
});

app.UseHangfireServer();

```

**Comment by pkunze** [comment]
Also, every Job Class/Method is affected. This is why I tend to believe there is a more general kind of problem going on.

**Comment by odinserj** [comment]
What if we try to throw an exception? It looks like there's another processing server that processes your jobs, and new code isn't running at all.

```csharp
public void SendEmail()
{
    throw new InvalidOperationException("Hello, world!");
}
```

**Comment by pkunze** [comment]
I just had the same Idea. ✌I will try this out and reply with results.

**Comment by pkunze** [comment]
I just created the following TestCase:

```csharp
[Route("api/[controller]")]
    [ApiController]
    public class HangfireDebuggingController : ControllerBase
    {
        private readonly IBackgroundJobClient queue;

        public HangfireDebuggingController(IBackgroundJobClient queue) => this.queue = queue;

        [HttpGet]
        [SwaggerOperation(Tags = new[] { "HangfireDebugging" }, OperationId = "EnqueueExceptionJob")]
        public void Get() => this.queue.Enqueue<HangfireDebuggingJob>(x => x.Execute());
    }

    public class HangfireDebuggingJob
    {
        public void Execute() => throw new Exception("Hello World!");
    }
```

Resulting in 
![image](https://user-images.githubusercontent.com/1416798/114878517-9d3ad800-9e00-11eb-94d4-db01d9fe2a84.png)

Both initial and after requeue.

**Comment by pkunze** [comment]
Interestingly enough, this is the first time I was able to reproduce this locally. IIS Express. Previously we only got this in QA (IIS)

**Comment by odinserj** [comment]
Really strange, `HangfireDebuggingJob` class was just created, right? What does the Servers page show?

**Comment by odinserj** [comment]
Do you have any global filters registered with `GlobalJobFilters.Filters.Add` method or `GlobalConfiguration.Configuration.UseFilter`?

**Comment by pkunze** [comment]
I will check & follow up as soon as I can

**Comment by pkunze** [comment]
We have a Job Filter which is registered in `Startup.Configure()`

```csharp
GlobalJobFilters.Filters.Add(new CanBePausedAttribute());
```

The Source code is:

```csharp
public class CanBePausedAttribute : JobFilterAttribute, IServerFilter
{
    public void OnPerforming(PerformingContext filterContext)
    {
        var values = filterContext.Connection.GetAllItemsFromSet("paused-jobs");

        // It's an ordinary job
        if (values.Contains(filterContext.BackgroundJob.Id))
        {
            filterContext.Canceled = true;
            return;
        }

        // Looking for recurring job chains
        foreach (var value in values)
        {
            var recurringJob = filterContext.Connection.GetRecurringJobs().FirstOrDefault(x => x.Job.Type.Name == value);

            if (recurringJob == null)
            {
                continue;
            }

            var jobDetails = JobStorage.Current.GetMonitoringApi().JobDetails(recurringJob.LastJobId);

            if (jobDetails != null && jobDetails.History.All(x => x.Reason != "Triggered using recurring job manager"))
            {
                filterContext.Canceled = true;
                return;
            }
        }
    }

    public void OnPerformed(PerformedContext filterContext)
    {
    }
}
```

**Comment by odinserj** [comment]
Please ensure your project doesn't have other filters that set the `filterContext.Canceled` property, because it is the only reason to have the mentioned behavior.

**Comment by pkunze** [comment]
Will check. Does this explain why the Jobs work when the DB is fresh?

**Comment by pkunze** [comment]
The Problem is the Filter above. The Issue can be closed. Thanks a lot for your Guidance! Is there any way I can send some beer over to you?

**Comment by odinserj** [comment]
You are welcome, actually this behavior will be changed in 1.8.0, and instead of Succeeded state in case of cancellation from a filter, we'll see the Deleted one with "Canceled by filter XXX" reason. So cases like this will be diagnosed much faster, and changes are already available in 1.8.0-beta2 released yesterday.
